File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import heapq

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.heap = []

        self.dfs(root, self.heap)
        for i in range(k - 1):
            heapq.heappop(self.heap)

        return heapq.heappop(self.heap)

    def dfs(self, root, heap):
        if not root:
            return 
        heapq.heappush(self.heap, root.val)
        self.dfs(root.left,heap)
        self.dfs(root.right, heap)
    #use a priority queue
    #pop until k

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int diameter;
    public int DiameterOfBinaryTree(TreeNode root) {
        diameter = 0;

        dfs(root);
        return diameter;
    }

    public int dfs(TreeNode root) {
        if (root is null)
            return 0;

        var left = dfs(root.left);
        var right = dfs(root.right);

        diameter = Math.Max(diameter, left + right);

        return 1 + Math.Max(left, right);
    }
}
